In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/Checkout.csv
/kaggle/input/Visits.csv
/kaggle/input/Cart.csv
/kaggle/input/Purchase.csv


In [4]:
import pandas as pd

In [8]:
visits = pd.read_csv('/kaggle/input/Visits.csv')

cart = pd.read_csv('/kaggle/input/Cart.csv')

checkout = pd.read_csv('/kaggle/input/Checkout.csv')

purchase = pd.read_csv('/kaggle/input/Purchase.csv')

print(visits.head())
print(cart.head())
print(checkout.head())
print(purchase.head())


                                user_id       visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188   4/7/2017 15:14
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  1/26/2017 14:24
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1   8/20/2017 8:23
3  6879527e-c5a6-4d14-b2da-50b85212b0ab  11/4/2017 18:15
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96  2/27/2017 11:25
                                user_id        cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168  11/7/2017 20:45
1  4397f73f-1da3-4ab3-91af-762792e25973   5/27/2017 1:35
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663   3/4/2017 10:38
3  b594862a-36c5-47d5-b818-6e9512b939b3   9/27/2017 8:22
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974  7/26/2017 15:48
                                user_id    checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d   6/25/2017 9:29
1  4ac186f0-9954-4fea-8a27-c081e428e34e   4/7/2017 20:11
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d  7/13/2017 11:38
3  89fe330a-8966-4756-8f7c-3bdbcd47279a  4/20/2017 16:15
4  3ccdaf69-2d30-40de-b083-5137

The data files are about a made-up company called Cool T-Shirts. The purpose is to analyze data on visits to the website and understand the funnel better.

In this case, The funnel is going to describe the following process:

1. A user visits CoolTShirts.com
2. A user adds a t-shirt to their cart
3. A user clicks “checkout”
4. A user actually purchases a t-shirt

In [14]:
visits_cart= pd.merge(visits, cart, how='left')

all_count=len(visits_cart)

null= visits_cart[visits_cart.cart_time.isnull()]

null_count=len(null)

no_cart= float(null_count)/all_count*100

print(no_cart)

80.50682261208577


Combined visits and cart using a left merge, and created a new dataframe called visits_cart to further understand the conversion from visiting the site to placing items in cart.The result shows 80.5% of users who visited Cool T-Shirts ended up not placing a t-shirt in their cart. 

In [15]:
cart_checkout= pd.merge(cart,checkout, how='left')

null2= cart_checkout[cart_checkout.checkout_time.isnull()]

all_count2= len(cart_checkout)
null_count2= len(null2)

no_checkout= float(null_count2)/all_count2*100

print(no_checkout)

20.930232558139537


Repeat the same process to merge cart and checkout files. 20.9% of users who placed items in their cart did not proceed to checkout. 

In [16]:
all_data= visits.merge(cart, how='left').merge(checkout, how='left').merge(purchase, how='left')

checkout_to_purchase= all_data[all_data.checkout_time.notnull()]

total_checkouts= len(checkout_to_purchase)

null3= checkout_to_purchase[checkout_to_purchase.purchase_time.isnull()]

total_non_purchase= len(null3)

no_purchase= float(total_non_purchase)/total_checkouts*100

print(no_purchase)

14.705882352941178


Continue to repeat the process and calculate the conversion from checkout to purchase. Only 14.7% of users who proceeded to checkout did not end up making purchases. Comparing all steps of the funnel,the pagevisit s the weakest with the highest percentage of users dropping out. 

In [25]:
all_data['time_to_purchase'] = int(all_data.purchase_time) - int(all_data.visit_time)

print(all_data.time_to_purchase)

print(all_data.time_to_purchase.mean())

SyntaxError: invalid syntax (<ipython-input-25-43455678aaba>, line 1)